File: week4_divid_and_conquer/BinarySearch/binary-search-python/binary_search.py
from sys import stdin, stdout

def binary_search_rec(A, left, right, key):
    # test for empty list
    if left > right:
        return -1

    # compare with middle element
    mid = (right + left) // 2

    if key == A[mid]:
        return mid
    elif key < A[mid]:
        # recursively search the lower half
        return binary_search_rec(A, left, mid-1, key)
    else:
        # recursively search the upper half
        return binary_search_rec(A, left, mid+1, key)

def binary_search(A, key):
    # Call recursive helper function
    return binary_search_rec(A, 0, len(A)-1, key)


if __name__ == '__main__':
    input = stdin.readline()
    data = list(map(int, input.split()))
    n = data[0]
    m = data[n + 1]
    a = data[1 : n+1]
    
    for x in data[n + 2:]:
        print(binary_search(a, x), end=" ")
    print()

File: week4_divid_and_conquer/BinarySearch/binary-search-cpp/binary_search.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int binarySearch(const vector<int> & a, int key);
int binarySearchRec(const vector<int> & a, int iLeft, int iRight, int key);

int main() {
    int n;
    cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); i++) {
	cin >> a[i];
    }
    int m;
    cin >> m;
    vector<int> b(m);
    for (int i = 0; i < m; ++i) {
	cin >> b[i];
    }
    for (int i = 0; i < m; ++i) {
	cout << binarySearch(a, b[i]) << ' ';
    }

    cout << endl;
    return 0;
}

// Search the array for the given key
// If found, return array index, otherwise, return -1
int binarySearch(const vector<int> & a, int key) {
    return binarySearchRec(a, 0, (int)a.size()-1, key);
}


// Recursive helper function for binarySearch
int binarySearchRec(const vector<int> & a, int iLeft, int iRight, int key) {

    // Test for empty list
    if (iLeft > iRight) return -1;

    // Compare with middle element
    int mid = (iRight + iLeft) / 2;   // truncate
    if (key == a[mid]) {
	return mid;
    } else if (key < a[mid]) {
	// Recursively search the lower half
	return binarySearchRec(a, iLeft, mid-1, key);
    } else {
	// Recursively search the upper half
	return binarySearchRec(a, mid+1, iRight, key);
    }
}
